tsl/src/fdw/deparse.c

```diff
@@ -834,6 +834,169 @@ deparseSelectStmtForRel(StringInfo buf, PlannerInfo *root, RelOptInfo *rel, List
 	deparseLockingClause(&context);
 }
 
+/*
+ * Construct "SELECT DISTINCT target_list" or "SELECT DISTINCT ON (col1, col..)
+ * target_list" statement to push down the DISTINCT clause to the remote side.
+ *
+ * We only allow references to basic "Vars" or constants in the DISTINCT exprs
+ *
+ * So, "SELECT DISTINCT col1" is fine but "SELECT DISTINCT 2*col1" is not.
+ *
+ * "SELECT DISTINCT col1, 'const1', NULL, col2" which is a mix of column
+ * references and constants is also supported. Everything else is not supported.
+ *
+ * Note that "SELECT DISTINCT col1, col2" is different from "SELECT DISTINCT
```

Comment:
Can you explain why those two are different without explicit ORDER BY there are no guarantees how values are returned and you would get all distinct combinations between col1 and col2 either way

Reply:
I have modified the `retrieved_attrs` to get the columns in the same order as is sent via the DISTINCT clause. So, when received, they are interpreted in the same sequence. Please look at `dc_varattnos` handling in the patch.

Reply:
I see that you are handling the 2 cases differently but why are you doing this? Why is dc_varattnos not a BitmapSet?

Reply:
A BitmapSet is ordered right? We want to retrieve the attributes in the order of their occurrence in the targetlist. Hence using an int List.

Reply:
Why do we want that? Lower scans usually dont bother with getting stuff in targetlist order and that adjustment happens in upper projection.

Reply:
Because `DISTINCT col1, col2` is different from `DISTINCT col2, col1` especially when ORDER BY is used, right?

Reply:
No it is the same and you need to use ORDER BY if you want specific order.Implicit ordering only happens when a Unique plan is chosen because Unique requires sorted input. Ordering requirements between nodes are verified with pathkeys and never assumed. If a Path does not have correct pathkey required for an operation a Sort node will be injected.

Reply:
The issue seems to be that 

```
SELECT DISTINCT device from hyper order by device;
```

is not sending the "order by device" in the pathkey because Postgres is choosing to use a Sort node as the top as I mentioned below. This seems to be the standard behavior as of now in regular tables with DISTINCT. What we will need is to ensure that we get a SkipScan like plan where the leaf node is the REMOTE SQL  containing the DISTINCT one instead of SkipScan.

Reply:
You can get rid of dc_varattnos completely since the required attributes are already tracked in attrs_used and rely on normal targetlist handling for the DISTINCT without ON and just handle DISTINCT ON in deparseDistinctClause

Reply:
Ok, thanks for confirming that SELECT DISTINCT order does not matter. Removed the `dc_varattnos` handling completely. The code is much simpler now than before. Handling targetlist only in `DISTINCT ON` case now.